### PR TITLE
#3062 Work on adding ability to convert between collections and arrays.

### PR DESCRIPTION
**Pull request by kag0**
So this is not ready for merge, has some outstanding issues. 
- tests - I made `ConvertAll` package private so I could slap some unit tests in there. What is the correct place for tests?
- API - as you can see from the test, when converting from a collection you have to use `ConvertAll` directly, since adding another `convert(Collection<?> collection, Class<? extends T[]> type)` method would conflict, and changing the existing one might not be what you want to do.

### CODE REVIEW

File: jOOQ/src/main/java/org/jooq/tools/Convert.java [hunk 1]
```diff
@@ -494,16 +494,31 @@ else if (fromClass == byte[].class) {
...
+                    Object[] fromArray = (Object[]) from;
+
+                    if (Collection.class.isAssignableFrom(toClass) &&
+                            toClass.isAssignableFrom(ArrayList.class)) {
+                        return (U) new ArrayList(Arrays.asList(fromArray));
+                    }
+                    else if (Collection.class.isAssignableFrom(toClass) &&
+                            toClass.isAssignableFrom(LinkedHashSet.class)) {
+                        return (U) new LinkedHashSet(Arrays.asList(fromArray));
+                    }
```
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/tools/Convert.java]: I like this implementation approach. Very cunning. Could you perhaps add a comment with an issue reference to the first `if`

```java
// [#3062] Default collections if no specific collection type was requested

File: jOOQ/src/main/java/org/jooq/tools/Convert.java [hunk 2]
```diff
@@ -430,7 +430,7 @@ private Convert() {}
     /**
      * The converter to convert them all.
      */
-    private static class ConvertAll<U> implements Converter<Object, U> {
+    static class ConvertAll<U> implements Converter<Object, U> {
```
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/tools/Convert.java]: I'd prefer not to open this up to package privacy. I understand you did this to access the class from tests, but this class shouldn't be accessible from tests. Tests should only call `Convert.xxx()` methods.

**Comment by lukaseder** [comment]
> tests - I made ConvertAll package private so I could slap some unit tests in there. What is the correct place for tests?

Our tests are not open source. Any place you put them is fine, but do note, I'll remove them again from the public repository after merging.

> API - as you can see from the test, when converting from a collection you have to use ConvertAll directly, since adding another convert(Collection<?> collection, Class<? extends T[]> type) method would conflict, and changing the existing one might not be what you want to do.

There's `convertArray()` for special cases. I'm happy with adding `convertCollection()` in a similar fashion

**Comment by kag0** [comment]
@lukaseder addressed those issues, let me know how it looks.

**Comment by lukaseder** [comment]
I'm currently postponing this. I'm not 100% sure what the side-effects of such an added feature are for it to be included in 3.9. Will look again into this later.

Sorry for all the delays.

File: jOOQ/src/main/java/org/jooq/tools/Convert.java [hunk 3]
```diff
@@ -517,7 +522,18 @@ else if (Collection.class.isAssignableFrom(toClass) &&
...
+                    Object[] list = f.stream()
+                            .map(e -> {
+                                if (!componentType.isAssignableFrom(e.getClass()))
+                                    return convert(e, componentType);
+                                return e;
+                            }).toArray();
+                    System.arraycopy(list, 0, dest, 0, dest.length);
+
+                    return (U) dest;
                 }
```
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/tools/Convert.java]: We can't generally make use of the `Stream` API in jOOQ as we still support Java 6. I will fix this code to retrofit it to our expectations.

**Comment by lukaseder** [comment]
Thanks again for your pull request. I've reviewed it again now. There's another issue that I've commented on, but I'll fix that on my side. Will merge now and commit against this issue.

File: jOOQ/src/main/java/org/jooq/tools/Convert.java (hunk 3 above)
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/tools/Convert.java]: I'm aware that Java 8 logic is scattered across this code (including references to `Optional`), but you don't see the full picture, unfortunately. The commercial distribution has special markers that delimit code with a Java 8 dependency and strips it from the Java 6 distribution. The Open Source edition doesn't have those markers.

Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/tools/Convert.java]: Another issue (although unlikely), the new code block might conflict with the existing block handling `byte[]` target types. Will switch order of branch...

**Pull request merged**